### PR TITLE
Add cluster-wide shard limit warnings

### PR DESCRIPTION
**Pull request by AthenaEryma**
In a future major version, we will be introducing a soft limit on the
number of shards in a cluster based on the number of nodes in the
cluster. This is intended to prevent operations which may 
unintentionally destabilize the cluster. 

This limit is configurable, and checked on operations which create or 
open shards and issue a warning if the operation would take the 
cluster over the configured limit.

There is an option to enable strict enforcement of the limit, which
turns the warnings into errors.  In a future release, the option will be
removed and strict enforcement will be the default (and only) behavior.

This PR will be followed by a 7.0-only PR which removes the enforcement
option and deprecation warnings and always enforces the limit.

Relates to #20705.

This is take 2 of #32856

### CODE REVIEW

File: docs/reference/modules/cluster/misc.asciidoc [hunk 1]
```diff
@@ -22,6 +22,48 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
 API can make the cluster read-write again.
 
 
+[[cluster-shard-limit]]
+
+==== Cluster Shard Limit
+
+In a Elasticsearch 7.0 and later, there will be a soft cap on the number of
```
Comment (colings86) [inline docs/reference/modules/cluster/misc.asciidoc]: Should we call it a "soft limit" to be in line with the terminology on similar settings elsewhere?

File: docs/reference/modules/cluster/misc.asciidoc [hunk 2]
```diff
@@ -22,6 +22,48 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
+
+If an operation, such as creating a new index, restoring a snapshot of an index,
+or opening a closed index would lead to the number of shards in the cluster
+going over this limit, the operation will issue a deprecation warning, or fail
+if `cluster.shards.enforce_max_per_node` is set.
+
+If the cluster is already over the cap, due to changes in node membership or
+setting changes, all operations that create or open indices will issue warnings
+or fail until either the cap is increased as described below, or some indices
+are closed or deleted to bring the number of shards below the cap.
```
Comment (colings86) [inline docs/reference/modules/cluster/misc.asciidoc]: As above I wonder if we should use "limit" instead of "cap"?

**Review by colings86 — COMMENTED** [review]
I left some documentation comments. The code seems good to me but it would be worth someone more familiar with this area of the code taking a look

File: docs/reference/modules/cluster/misc.asciidoc (hunk 1 above)
Comment (gwbrown) [reply docs/reference/modules/cluster/misc.asciidoc]: Yes, I'll change all the instances of "cap" to "limit" - thanks!

File: docs/reference/modules/cluster/misc.asciidoc [hunk 3]
```diff
@@ -22,6 +22,48 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
+intended to prevent operations which may unintentionally destabilize the
+cluster. Until 7.0, actions which would result in the cluster going over the
+limit will issue a deprecation warning. Strict enforcement of this behavior can
+be enabled by setting the cluster setting `cluster.shards.enforce_max_per_node`
+to `true`.
+
+NOTE: `cluster.shards.enforce_max_per_node` cannot be set to `false`, as this
+setting will be removed in 7.0 and the limit will always be enforced. To return
+to the default behavior for your Elasticsearch version, set this setting to
+`"default"`.
```
[the remark this answers is not in the document]
Comment (gwbrown) [reply docs/reference/modules/cluster/misc.asciidoc]: I did this based on a conversation with @jasontedor a while ago, where he made very similar comments, but I think I misunderstood what he was suggesting at the time.  I'll reevaluate this setting and change it as appropriate.

[the remark this answers is not in the document]
Comment (gwbrown) [reply docs/reference/modules/cluster/misc.asciidoc]: Just talked with @jasontedor again - this setting is going to go away and become a system property, which can only be unset or `true`.

File: docs/reference/modules/cluster/misc.asciidoc [hunk 4]
```diff
@@ -22,6 +22,47 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
 
 
+[[cluster-shard-limit]]
+
+==== Cluster Shard Limit
+
+In a Elasticsearch 7.0 and later, there will be a soft limit on the number of
+shards in a cluster, based on the number of nodes in the cluster.  This is
+intended to prevent operations which may unintentionally destabilize the
+cluster. Until 7.0, actions which would result in the cluster going over the
```
Comment (jasontedor) [inline docs/reference/modules/cluster/misc.asciidoc]: `Until` -> `Prior to`

File: docs/reference/modules/cluster/misc.asciidoc [hunk 5]
```diff
@@ -22,6 +22,47 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
+limit will issue a deprecation warning.
+
+NOTE: You can set the system property `es.enforce.shard_limit` to `true` to opt
```
Comment (jasontedor) [inline docs/reference/modules/cluster/misc.asciidoc]: I don't think we need to namespace this under `enforce`, so `es.enforce.shard_limit` -> `es.enforce_shard_limit`. And perhaps it should more closely reflect the name of the setting, so `es.enforce.shard_limit` -> `es.enforce_max_shards_per_node`.

File: docs/reference/modules/cluster/misc.asciidoc [hunk 6]
```diff
@@ -22,6 +22,47 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
+only behavior.
+
+If an operation, such as creating a new index, restoring a snapshot of an index,
+or opening a closed index would lead to the number of shards in the cluster
+going over this limit, the operation will issue a deprecation warning.
+
+If the cluster is already over the limit, due to changes in node membership or
+setting changes, all operations that create or open indices will issue warnings
+until either the limit is increased as described below, or some indices
+are closed or deleted to bring the number of shards below the limit.
```
Comment (jasontedor) [inline docs/reference/modules/cluster/misc.asciidoc]: Would you link to the sections of the documentation relevant to closing, and separately deleting an index?

File: docs/reference/modules/cluster/misc.asciidoc [hunk 7]
```diff
@@ -22,6 +22,47 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
+
+Replicas count towards this limit, but closed indexes do not. An index with 5
+primary shards and 2 replicas will be counted as 15 shards.  Any closed index
+is counted as 0, no matter how many shards and replicas it contains.
+
+The limit defaults to 1,000 shards per node, and be dynamically adjusted using
+the following property:
+
+`cluster.shards.max_per_node`::
```
Comment (jasontedor) [inline docs/reference/modules/cluster/misc.asciidoc]: I am doubting whether this needs to be in a `shards` namespace. How about `cluster.shards.max_per_node` -> `cluster.max_shards_per_node`.

File: docs/reference/migration/migrate_7_0/cluster.asciidoc [hunk 8]
```diff
@@ -14,3 +14,9 @@ primary shards of the opened index to be allocated.
 
 ==== Shard preferences `_primary`, `_primary_first`, `_replica`, and `_replica_first` are removed
 These shard preferences are removed in favour of the `_prefer_nodes` and `_only_nodes` preferences.
+
+==== Cluster-wide shard soft limit
+Clusters now have soft limits on the total number of open shards in the cluster
+based on the number of nodes and the `cluster.shards.max_per_node` cluster
```
Comment (jasontedor) [inline docs/reference/migration/migrate_7_0/cluster.asciidoc]: How about `cluster.shards.max_per_node` -> cluster.max_shards_per_node`.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 9]
```diff
@@ -124,9 +123,11 @@
     public interface Custom extends NamedDiffable<Custom>, ToXContentFragment, ClusterState.FeatureAware {
 
         EnumSet<XContentContext> context();
-
     }
 
+    public static final Setting<Integer> SETTING_CLUSTER_MAX_SHARDS_PER_NODE =
+        Setting.intSetting("cluster.shards.max_per_node", 1000, 1, Property.Dynamic, Property.NodeScope);
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: `cluster.shards.max_per_node` -> `cluster.max_shards_per_node`.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 10]
```diff
@@ -676,10 +683,29 @@ public IndexGraveyard indexGraveyard() {
...
+     * Gets the total number of shards from all indices, including replicas and
+     * closed indices.
+     * @return The total number shards from all indices.
+     */
     public int getTotalNumberOfShards() {
         return this.totalNumberOfShards;
     }
 
+    /**
+     * Gets the total number of active shards from all indices. Includes
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Not all open indices are active. Let us be precise here and say `open`.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 11]
```diff
@@ -676,10 +683,29 @@ public IndexGraveyard indexGraveyard() {
...
+     * replicas, but does not include shards that are part of closed indices.
+     * @return The total number of active shards from all indices.
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: `active` -> `open`

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 12]
```diff
@@ -587,19 +589,29 @@ public void onFailure(String source, Exception e) {
 
     private void validate(CreateIndexClusterStateUpdateRequest request, ClusterState state) {
         validateIndexName(request.index(), state);
-        validateIndexSettings(request.index(), request.settings(), forbidPrivateIndexSettings);
+        validateIndexSettings(request.index(), request.settings(), state, forbidPrivateIndexSettings);
     }
 
-    public void validateIndexSettings(
-            final String indexName, final Settings settings, final boolean forbidPrivateIndexSettings) throws IndexCreationException {
+    public void validateIndexSettings(String indexName, final Settings settings, ClusterState clusterState, final boolean forbidPrivateIndexSettings) throws IndexCreationException {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: `clusterState` can be final.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java [hunk 13]
```diff
@@ -217,4 +222,25 @@ public ClusterState execute(ClusterState currentState) {
...
+        Optional<String> error = IndicesService.checkShardLimit(shardsToOpen, currentState, deprecationLogger);
+        if (error.isPresent()) {
+            ValidationException ex = new ValidationException();
+            ex.addValidationError(error.get());
+            throw ex;
+        }
+
+    }
+
+
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java]: There is an extraneous newline here.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java [hunk 14]
```diff
@@ -217,4 +222,25 @@ public ClusterState execute(ClusterState currentState) {
         });
     }
 
+    static void validateShardLimit(ClusterState currentState, Index[] indices, DeprecationLogger deprecationLogger) {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java]: Javadocs please.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 15]
```diff
@@ -587,19 +589,29 @@ public void onFailure(String source, Exception e) {
...
+        shardAllocation.ifPresent(validationErrors::add);
+
         if (validationErrors.isEmpty() == false) {
             ValidationException validationException = new ValidationException();
             validationException.addValidationErrors(validationErrors);
             throw new IndexCreationException(indexName, validationException);
         }
     }
 
+    static Optional<String> checkShardLimit(Settings settings, ClusterState clusterState, DeprecationLogger deprecationLogger) {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: Javadocs please.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 16]
```diff
@@ -156,6 +158,20 @@
     public static final String INDICES_SHARDS_CLOSED_TIMEOUT = "indices.shards_closed_timeout";
     public static final Setting<TimeValue> INDICES_CACHE_CLEAN_INTERVAL_SETTING =
         Setting.positiveTimeSetting("indices.cache.cleanup_interval", TimeValue.timeValueMinutes(1), Property.NodeScope);
+    private static final boolean ENFORCE_SHARD_LIMIT;
+    static {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: A newline would be appreciated here. 😄

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 17]
```diff
@@ -156,6 +158,20 @@
...
+        final String ENFORCE_SHARD_LIMIT_KEY = "es.enforce.shard_limit";
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: `es.enforce_shard_limit` -> `es.enforce_max_shards_per_node`.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 18]
```diff
@@ -156,6 +158,20 @@
...
+        final String enforceShardLimitSetting = System.getProperty(ENFORCE_SHARD_LIMIT_KEY);
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: `enforceShardLimitSetting` ->` enforceMaxShardsPerNode`

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 19]
```diff
@@ -156,6 +158,20 @@
     public static final String INDICES_SHARDS_CLOSED_TIMEOUT = "indices.shards_closed_timeout";
     public static final Setting<TimeValue> INDICES_CACHE_CLEAN_INTERVAL_SETTING =
         Setting.positiveTimeSetting("indices.cache.cleanup_interval", TimeValue.timeValueMinutes(1), Property.NodeScope);
+    private static final boolean ENFORCE_SHARD_LIMIT;
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: `ENFORCE_SHARD_LIMIT` -> `ENFORCE_MAX_SHARDS_PER_NODE`.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 20]
```diff
@@ -1347,4 +1363,42 @@ public void clearIndexShardCache(ShardId shardId, boolean queryCache, boolean fi
     public boolean isMetaDataField(String field) {
         return mapperRegistry.isMetaDataField(field);
     }
+
+    /**
+     * Checks to see if an operation can be performed without taking the cluster
+     * over the cluster-wide shard limit. Adds a deprecation warning or returns
+     * an error message as appropriate
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: You don't have to wrap these so narrowly, we can use the full 140-column line length here.

**Review by jasontedor — CHANGES_REQUESTED** [review]
This is looking good, I left some comments, mostly about changing the keys.

**Comment by gwbrown** [comment]
Thanks! I've addressed your comments, can you re-review @jasontedor?

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by gwbrown** [comment]
Thanks for the reviews!

**Pull request merged**